ompi/request/grequest.c

```diff
@@ -214,13 +236,47 @@ int ompi_grequest_invoke_query(ompi_request_t *request,
         if (g->greq_funcs_are_c) {
             rc = g->greq_query.c_query(g->greq_state, status);
         } else {
+            /* request->req_status.MPI_ERROR was initialized to success
+             * and it's meant to be unmodified in the case of callback
+             * success, and set when callbacks return a failure.  But
+             * if we leave fstatus uninitialized this sets
+             * req_status.MPI_ERROR to whatever happened to be on the
+             * stack at fstatus (f_query isn't supposed to directly set
+             * its status.MPI_ERROR, according to the standard)
+             *
+             * So the Status_c2f below only really cares about transferring
+             * the MPI_ERROR setting into fstatus so that when it's transferred
+             * back in the f2c call, it has the starting value.
+             */
             MPI_Fint ierr;
             MPI_Fint fstatus[sizeof(MPI_Status) / sizeof(int)];
+            MPI_Status_c2f(status, fstatus);
             g->greq_query.f_query((MPI_Aint*)g->greq_state, fstatus, &ierr);
             MPI_Status_f2c(fstatus, status);
             rc = OMPI_FINT_2_INT(ierr);
         }
     }
+    /* When called from Wait/Test* I had a concern about this part
+     * of the 3.1 standard:
+     *  >  All callback functions return an error code. The code is
+     *  >  passed back and dealt with as appropriate for the error
+     *  >  code by the MPI function that invoked the callback function.
+     *  >  For example, if error codes are returned then the error code
+     *  >  returned by the callback function will be returned by the
+     *  >  MPI function that invoked the callback function. In the
+     *  >  case of an MPI_{WAIT|TEST}{ANY} call that invokes both
+     *  >  query_fn and free_fn, the MPI call will return the error
+     *  >  code returned by the last callback, namely free_fn.
+     * which sounds a bit like we should ignore query_fn()'s return
+     * for Wait/Test*, but my current interpretation is that the clauses
+     * "as appropriate" and "In the case of ... that invokes both"
+     * justifies our current behavior, where a failure from query_fn()
+     * makes it way into the MPI return or into status[i].MPI_ERROR
+     * and results in free_fn() not being called for that request.
+     *
+     * That seems like the most reasonable behavior, and I think it's
+     * justifiable from the standard too.
+     */
```

Comment:
page 637 in MPI 4.0 states that for any completing call (wait and test)
 > if the MPI call was passed an array of statuses, then MPI will return in each of the statuses that correspond to a completed generalized request the error code returned by the corresponding invocation of its free_fn callback function.